src/util/compat.hpp

```diff
@@ -34,4 +34,27 @@ inline void compiler_barrier_if_xlc_leq(unsigned ver) {
 template <typename X>
 inline constexpr bool isinf(X x) { return std::isinf(x); }
 
+// Work around a bad inlining-related optimization with icpc 16.0.3 and -xMIC-AVX512,
+// by forcing a computation.
+
+template <typename X>
+inline void sink(const X& x) {
```

Comment:
So, I think that the question here is why cast to `char*` instead of using `X` directly as follows:
```
template <typename X>
inline void sink(const X& x) {
    X buf;
    volatile X* bufptr = &buf;
    bufptr[0] = x;
}
```

The main reason I can see is that this requires `X` to be copyable, which might not always be the case. By casting to `char*` Sam is forcing the compiler to copy the bytes. Maybe overkill because we use this in just one spot to sink a `std::pair`, which is copyable. But it is more generic.